design/design-doc-composition.md

```diff
@@ -0,0 +1,1079 @@
+        reclaimPolicy: Delete
+    patches:
+    - fromFieldPath: "metadata.uid"
+      toFieldPath: "spec.writeConnectionSecretToRef.name"
+    - fromFieldPath: "spec.engineVersion"
+      toFieldPath: "spec.forProvider.version"
+    - fromFieldPath: "spec.storageGB"
+      toFieldPath: "spec.forProvider.storageMB"
+      transforms:
+      - type: math
+        math:
+          multiply: 1024
+    - fromFieldPath: "spec.region"
+      toFieldPath: "spec.forProvider.location"
+      transforms:
+      - type: map
+        map:
+          us-west: "West US"
+          us-east: "East US"
+    # Specifies the (potentially sensitive) connection details that this 'to'
+    # resource should expose to the 'from' resource. Names are unique across all
+    # 'to' resources within this composition. Ignored by application resources.
+    connectionDetails:
+    - name: username
+      fromConnectionSecretKey: username
+    - name: password
+      fromConnectionSecretKey: password
+    - name: endpoint
+      fromConnectionSecretKey: endpoint
+  - base:
+      apiVersion: database.azure.crossplane.io/v1alpha3
+      kind: MySQLServerVirtualNetworkRule
+      spec:
+        name: my-cool-vnet-rule
+        serverNameSelector:
+          matchComposite: true
+        resourceGroupNameSelector:
+          matchComposite: true
+        properties:
+          virtualNetworkSubnetIdRef:
+            name: sample-subnet
+        reclaimPolicy: Delete
+        providerRef:
+          name: azure-provider
+```
+
+This document proposes compositions be immutable. This property would allow
+updates to a 'from' resource to be propagated to the various 'to' resources but
+keep the set of 'to' resources stable over the lifetime of the 'from' resource,
+allowing controllers to avoid complex update and garbage collection logic.
+
+### ApplicationDefinition
+
+A `kind: ApplicationDefinition` defines a new kind of custom resource that
+represents an application - for example a `kind: Wordpress` custom resource. An
+application resource is namespaced, and may only compose resources within its
+namespace. Applications are typically defined by an _application developer_ -
+the developer of Wordpress would author the `kind: ApplicationDefinition` that
+defines the schema of a `kind: Wordpress` resource. Application resources are
+typically authored by an _application operator_.
+
+Here's an example that defines a new application resource of `kind: Wordpress`:
+
+```yaml
+apiVersion: crossplane.io/v1alpha1
+kind: ApplicationDefinition
+metadata:
+  # ApplicationDefinition names are subject to the constraints of Kubernetes
+  # CustomResourceDefinition names. They must be of the form <plural>.<group>.
+  name: wordpresses.apps.example.org
+spec:
+  # Defines the structural schema and GroupVersionKind of this application. Only
+  # a single API version of the application may exist. Additional fields will be
+  # injected to support composition machinery.
+  crdSpecTemplate:
```

Comment:
I'd suggest only expose openAPIV3Schema, or even more simpler UX here (like name:type pair).

`crdSpecTemplate` requires too much for a app developer.

Reply:
I do think there's an opportunity to converge this work with OAM at some point - we should discuss. The key point with a `Composition` here is that it's a unit of reuse at the API level -- i.e. I can create a group of components and wrap them in a new type with a well defined schema. AppConfigs in OAM are schemaless right now.

Let's definitely explore this further.